streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/DefaultTaskInfo.java

```diff
@@ -0,0 +1,120 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.internals.assignment;
+
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Collections.unmodifiableSet;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.streams.kstream.ValueMapper;
+import org.apache.kafka.streams.processor.TaskId;
+import org.apache.kafka.streams.processor.assignment.TaskInfo;
+
+public class DefaultTaskInfo implements TaskInfo {
+
+    private final TaskId id;
+    private final boolean isStateful;
+    private final Map<TopicPartition, Set<String>> partitionToRackIds;
+    private final Set<String> stateStoreNames;
+    private final Set<TopicPartition> inputTopicPartitions;
+    private final Set<TopicPartition> changelogTopicPartitions;
+
+    public DefaultTaskInfo(final TaskId id,
+                           final boolean isStateful,
+                           final Map<TopicPartition, Set<String>> partitionToRackIds,
+                           final Set<String> stateStoreNames,
+                           final Set<TopicPartition> inputTopicPartitions,
+                           final Set<TopicPartition> changelogTopicPartitions) {
+        this.id = id;
+        this.partitionToRackIds = unmodifiableMap(partitionToRackIds);
+        this.isStateful = isStateful;
+        this.stateStoreNames = unmodifiableSet(stateStoreNames);
+        this.inputTopicPartitions = unmodifiableSet(inputTopicPartitions);
+        this.changelogTopicPartitions = unmodifiableSet(changelogTopicPartitions);
+    }
+
+    public static DefaultTaskInfo of(final TaskId taskId,
+                                     final boolean isStateful,
+                                     final ValueMapper<TopicPartition, String> previousOwnerForPartition,
+                                     final Map<String, Optional<String>> rackForConsumer,
+                                     final Map<TaskId, Set<TopicPartition>> inputPartitionsForTask,
+                                     final Map<TaskId, Set<TopicPartition>> changelogPartitionsForTask) {
+
+        final Set<TopicPartition> inputPartitions = inputPartitionsForTask.get(taskId);
+        final Set<TopicPartition> changelogPartitions = changelogPartitionsForTask.get(taskId);
+        final Map<TopicPartition, Set<String>> racksForPartition = new HashMap<>();
+
+        inputPartitions.forEach(partition -> {
+            racksForPartition.computeIfAbsent(partition, k -> new HashSet<>());
+            final String consumer = previousOwnerForPartition.apply(partition);
+            final Optional<String> rack = rackForConsumer.get(consumer);
+            rack.ifPresent(s -> racksForPartition.get(partition).add(s));
+        });
```

Comment:
Ah, this is not computing the right rack id -- this would be the rack id of the KafkaStreams node that had this partition assigned during the last rebalance. What we want is the rack.id of the broker node(s) that host this partition. This is going to be a bit complex so let's chat online (ditto for the changelogPartitions rack info as well)